src/plugins/intel_gpu/src/kernel_selector/kernels/fully_connected/fully_connected_kernel_bf_tiled.cpp

```diff
@@ -522,29 +577,53 @@ void FullyConnected_bf_tiled::GetUpdateDispatchDataFunc(KernelData& kd) const {
+            // - kd.kernels[N-1] for batches <= 240 (default version)
+            // - kd.kernels[N] for batches >= 256 (slm version)
             const auto default_alignment = 16;
-            // We can use SLM version if `output_batch + default_alignment > 256` because memory and batch are aligned (whether 16 or 64 elements)
-            const auto skip_kernel_idx = output_batch + default_alignment > 256 ? 0 : 1;
-            const auto execute_kernel_idx = 1 - skip_kernel_idx;
+            // We can use SLM version if `output_batch + default_alignment > min_slm_size(256)` because memory and batch are aligned (whether 16 or 64 elements)
+            const auto execute_type = (output_batch + default_alignment > min_slm_size) ? KernelType::SLM : KernelType::DEFAULT;
+            const auto execute_kernel_idx = ((execute_type == KernelType::SLM) ? 1 : 0) + kernel_offset;
+            const auto skip_kernel_idx = ((execute_type == KernelType::SLM) ? 0 : 1) + kernel_offset;
+
 
+            // Check default or SLM version FC, and disable remain version
             kd.kernels[skip_kernel_idx].skip_execution = true;
 
-            GPU_DEBUG_TRACE_DETAIL << "FC bf tiled: " << (execute_kernel_idx == 1 ? "SLM" : "Default") << " shape-agnostic kernel version "
+            GPU_DEBUG_TRACE_DETAIL << "FC bf tiled: " << (execute_type == KernelType::SLM ? "SLM" : "Default") << " shape-agnostic kernel version "
                                     << "will be used for batch size = " << output_batch << "\n";
 
-            auto dispatchData = SetDefault(prim_params, -1, execute_kernel_idx);
+            auto dispatchData = SetDefault(prim_params, -1, static_cast<int>(execute_type));
             kd.kernels[execute_kernel_idx].params.workGroups.global = dispatchData.gws;
             kd.kernels[execute_kernel_idx].params.workGroups.local = dispatchData.lws;
             kd.kernels[execute_kernel_idx].skip_execution = KernelData::SkipKernelExecution(prim_params);
+
+            if (!kd.internalBufferSizes.empty()) {
+                // Pre-quantizing kernel was generated. Update the kernel and intermediate buffers or disable it.
+                if (execute_type == KernelType::DEFAULT) {
+                    kd.kernels[0].skip_execution = true;
+                } else {
+                    kd.kernels[0].skip_execution = false;
+                    size_t input_f = get_input_bf_size(prim_params).second;
+                    size_t input_size = input_f * dispatchData.tile_m * dispatchData.gws[2];
+
+                    if (kd.kernels[0].params.workGroups.global[0] < (input_size / quantize_grp_size)) {
+                        kd.internalBufferSizes.clear();
+                        kd.internalBufferSizes.push_back(input_size);                           // quantized input is char type
+                        kd.internalBufferSizes.push_back(input_size / quantize_grp_size * 2);   // de_quan_scale is half type
```

Comment:
how does it interact with preallocation? For example, if prompt length changes from 1000, 1001, 1002, 1003 ... does preallocation happen properly?

Reply:
For dynamic, out_b would be 0 for pre-allocation. prompt length would be referred at runtime here. And input_size would be aligned by dispatchData.tile_m. So it would be fine.